core/common/src/main/java/alluxio/shell/ShellCommand.java

```diff
@@ -0,0 +1,145 @@
+   *
+   * @return the output
+   * @throws IOException if the command returns a non-zero exit code
+   */
+  public String run() throws IOException {
+    Process process = new ProcessBuilder(mCommand).redirectErrorStream(true).start();
+
+    BufferedReader inReader =
+            new BufferedReader(new InputStreamReader(process.getInputStream(),
+                    Charset.defaultCharset()));
+
+    try {
+      // read the output of the command
+      StringBuilder output = new StringBuilder();
+      String line = inReader.readLine();
+      while (line != null) {
+        output.append(line);
+        output.append("\n");
+        line = inReader.readLine();
+      }
+      // wait for the process to finish and check the exit code
+      int exitCode = process.waitFor();
+      if (exitCode != 0) {
+        throw new ShellUtils.ExitCodeException(exitCode, output.toString());
+      }
+      return output.toString();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new IOException(e);
+    } finally {
+      // close the input stream
+      try {
+        // JDK 7 tries to automatically drain the input streams for us
+        // when the process exits, but since close is not synchronized,
+        // it creates a race if we close the stream first and the same
+        // fd is recycled. the stream draining thread will attempt to
+        // drain that fd!! it may block, OOM, or cause bizarre behavior
+        // see: https://bugs.openjdk.java.net/browse/JDK-8024521
+        // issue is fixed in build 7u60
+        InputStream stdout = process.getInputStream();
+        synchronized (stdout) {
+          inReader.close();
+        }
+      } catch (IOException e) {
+        LOG.warn("Error while closing the input stream", e);
+      }
+      process.destroy();
+    }
+  }
+
+  /**
+   * Runs a command and returns its output and exit code.
+   * No matter it succeeds or not.
+   * Stderr is redirected to stdout.
+   *
+   * @return {@link CommandReturn} object representation of stdout, stderr and exit code
+   */
+  public CommandReturn runTolerateFailure() throws IOException {
+    Process process = new ProcessBuilder(mCommand).redirectErrorStream(true).start();
+    CommandReturn cr = null;
+
+    try (BufferedReader inReader =
+                 new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+      // read the output of the command
+      StringBuilder stdout = new StringBuilder();
+      String outLine = inReader.readLine();
+      while (outLine != null) {
+        stdout.append(outLine);
+        stdout.append("\n");
+        outLine = inReader.readLine();
+      }
+
+      // wait for the process to finish and check the exit code
+      int exitCode = process.waitFor();
+      if (exitCode != 0) {
+        // log error instead of throwing exception
+        LOG.warn(String.format("Non-zero exit code from command %s%nExit Code: %s",
```

Comment:
Can we keep the error code in the same line, something like "Non-zero exit code (%d) from command: %s"